scripts/ci/dag.py

```diff
@@ -0,0 +1,169 @@
+
+    def walk(self, f: Callable[[_T], None], max_tokens: int, refill_interval_s: float) -> None:
+        """
+        Process the graph in parallel.
+
+        Each node in the graph is processed only once all of its dependencies have been processed.
+
+        Concurrency is limited by the following bucket rate limiting algorithm:
+        * Processing may not begin until a token can be acquired from the bucket.
+        * There are at most `max_tokens - in_progress_tasks` in the bucket at any time.
+        * Tokens are refreshed every `refill_interval_s`.
+        """
+
+        num_cpus = cpu_count()
+        # we need one main thread + N worker threads, so `N = num_cpus - 1`,
+        # and we need at least 1 worker thread.
+        num_workers = num_cpus - 1 if num_cpus > 0 else 1
+        with ThreadPoolExecutor(max_workers=num_workers) as p:
+            task_queue: Queue[_T] = Queue()
+            done_queue: Queue[_T] = Queue()
+            shutdown: EventClass = Event()
+
+            def worker(n: int) -> None:
+                # Attempt to grab a task from the queue,
+                # execute it, then put it in the done queue.
+                while not shutdown.is_set():
+                    try:
+                        node = task_queue.get_nowait()
+                        f(node)
+                        done_queue.put(node)
+                    except Empty:
+                        time.sleep(0)  # yield to prevent busy-looping
+                        continue
+
+            for n in range(0, num_workers):  # start all workers
+                p.submit(worker, n)
+
+            tokens = max_tokens
+            in_progress = 0
+            last_refill = time.time()
+            while not self._is_done():
+                # This loop has two parts, `push` and `pull`.
+                #
+                # The `push` loop attempts to push tasks
+                # onto the `task_queue` while there are
+                # some tasks ready to go, and some tokens left.
+                # It is also responsible for refreshing the bucket.
+                #
+                # The `pull` loop attempts to retrieve done
+                # tasks and decrement the dependency counter on
+                # their dependents.
+                #
+                # Once a node has no pending dependencies left,
+                # it becomes ready and will be queued in one of
+                # the iterations of the `push` loop.
+                #
+                # It's important to always use non-blocking `get`
+                # with the task and done queues.
+                # This is so that both the push and pull loops
+                # can eventually make progress.
+                while len(self._queue) > 0:  # push loop
+                    now = time.time()
+                    if now - last_refill > refill_interval_s:
+                        tokens = max_tokens - in_progress
+                        last_refill = now
```

Comment:
I would expect `tokens += 1` here, i.e. trickle in just one more token at a time

Reply:
Why is that? Usually in the bucket rate limiting schemes I've seen, the bucket is always refilled to full capacity. Here I'm subtracting `num_in_progress` so that we don't run into a situation where there are more than `max_tokens` tasks running in parallel.

Reply:
It depends on the purpose of `refill_interval_sec` I suppose. If the goal is rate-limiting so that we start at most one job per `refill_interval_sec` seconds, then we should do something like

```
while now - last_refill > refill_interval_sec:
    tokens += 1
    last_refill += refill_interval_sec
```

…but rate-limiting ourselves is actually more easily handled in the job, i.e. in `publish_crates` with a gobal:

```py
last_publish_time = -10000

def publish_crate(…):
    min_sec_between_publish = 1 # to not get rate-limited by crates.io
    sec_since_last_publish = time() - last_publish_time
    sleep(max(0, min_sec_between_publish - sec_since_last_publish))

    # publish here
    last_publish_time = time()

Reply:
> If the goal is rate-limiting so that we start at most one job per refill_interval_sec seconds <...>

The goal is to ensure that we never publish more than 30 crates per minute (that's the rate limit in crates.io), but to also not unnecessarily slow down the publishing process. We want to spawn as many publishes in parallel as we can.

This made me realise that `refill_interval_sec` should be set to `60` 😅.